### ISSUE TITLE
Integrate Clang Static Analyzer

### ISSUE THREAD

**Issue by guwirth** [SonarOpenCommunity/sonar-cxx#547]
The Clang Static Analyzer is a source code analysis tool that finds bugs in C, C++, and Objective-C programs (http://clang-analyzer.llvm.org). 

Support this tool like Cppcheck.

**Comment by mjdetullio** [comment]
I started an implementation of this on my fork of the sonar-objective-c plugin if you want to kang it.  It's designed to read from the multi-plist output format that xcodebuild produces when it executes Clang.  The individual rules aren't defined yet, so at the moment it clumps all results under one rule and sets the issue message in the format of "<clang category> - <rule message>".

https://github.com/mjdetullio/sonar-objective-c/commit/d7426ffd7f5d277dde13544da04f67b020055441

**Comment by guwirth** [comment]
> I started an implementation of this on my fork of the sonar-objective-c plugin if you want to kang it. It's designed to read from the multi-plist output format that xcodebuild produces when it executes Clang.

@mjdetullio yes for sure that is interesting for us. 

>  The individual rules aren't defined yet, 

Do you know if there is a possibility to create a warning list with clang?  Some other tool has such command line options (e.g. CppCheck).

**Comment by mjdetullio** [comment]
According to the docs (http://clang-analyzer.llvm.org/checker_dev_manual.html) you can run ```clang -cc1 -analyzer-checker-help```, however the plist output does not contain the Clang keys or descriptions from that command's output on each of the issues.  I'm not sure if other formats will include the checker keys since as far as I know xcodebuild only produces multi-plist format.

For example, that command will list this as a checker:

```osx.cocoa.AtSync                Check for nil pointers used as mutexes for @synchronized```

But in a report that contains that issue type, neither key is found, as you can see from the below sample plist output (bottom is the relevant part):

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
    <dict>
        <key>clang_version</key>
        <string>Apple LLVM version 7.0.0 (clang-700.0.72)</string>
        <key>files</key>
        <array>
            <string>/absolute/path/to/DateCalculation.m</string>
        </array>
        <key>diagnostics</key>
        <array>
            <dict>
                <key>path</key>
                <array>
                    <dict>
                        <key>kind</key><string>control</string>
                        <key>edges</key>
                        <array>
                            <dict>
                                <key>start</key>
                                <array>
                                    <dict>
                                        <key>line</key><integer>349</integer>
                                        <key>col</key><integer>5</integer>
                                        <key>file</key><integer>0</integer>
                                    </dict>
                                    <dict>
                                        <key>line</key><integer>349</integer>
                                        <key>col</key><integer>10</integer>
                                        <key>file</key><integer>0</integer>
                                    </dict>
                                </array>
                                <key>end</key>
                                <array>
                                    <dict>
                                        <key>line</key><integer>351</integer>
                                        <key>col</key><integer>5</integer>
                                        <key>file</key><integer>0</integer>
                                    </dict>
                                    <dict>
                                        <key>line</key><integer>351</integer>
                                        <key>col</key><integer>8</integer>
                                        <key>file</key><integer>0</integer>
                                    </dict>
                                </array>
                            </dict>
                        </array>
                    </dict>
                    <!-- additional path entries omitted -->
                    <dict>
                        <key>kind</key><string>event</string>
                        <key>location</key>
                        <dict>
                            <key>line</key><integer>290</integer>
                            <key>col</key><integer>5</integer>
                            <key>file</key><integer>0</integer>
                        </dict>
                        <key>ranges</key>
                        <array>
                            <array>
                                <dict>
                                    <key>line</key><integer>290</integer>
                                    <key>col</key><integer>5</integer>
                                    <key>file</key><integer>0</integer>
                                </dict>
                                <dict>
                                    <key>line</key><integer>295</integer>
                                    <key>col</key><integer>2</integer>
                                    <key>file</key><integer>0</integer>
                                </dict>
                            </array>
                        </array>
                        <key>depth</key><integer>1</integer>
                        <key>extended_message</key>
                        <string>Nil value used as mutex for @synchronized() (no synchronization will occur)</string>
                        <key>message</key>
                        <string>Nil value used as mutex for @synchronized() (no synchronization will occur)</string>
                    </dict>
                </array>
                <key>description</key><string>Nil value used as mutex for @synchronized() (no synchronization will occur)</string>
                <key>category</key><string>Logic error</string>
                <key>type</key><string>Nil value used as mutex for @synchronized() (no synchronization will occur)</string>
                <key>issue_context_kind</key><string>Objective-C method</string>
                <key>issue_context</key><string>dateComponentsForDate:inCalendar:</string>
                <key>issue_hash</key><string>5</string>
                <key>location</key>
                <dict>
                    <key>line</key><integer>290</integer>
                    <key>col</key><integer>5</integer>
                    <key>file</key><integer>0</integer>
                </dict>
            </dict>
        </array>
    </dict>
</plist>
```

**Comment by mjdetullio** [comment]
Had to look this up in the source (https://github.com/llvm-mirror/clang/blob/master/include/clang/StaticAnalyzer/Core/Analyses.def), but the values for the ```-analyzer-output=``` option are:

```
html
plist
plist-multi-file
plist-html
text
```

None of these reports include the key or description from the help output :(

**Comment by guwirth** [comment]
no contributors, close this

**Comment by gyorb** [comment]
Hi,

I'm working on this feature. There is a prototype on my fork which can parse the plist files generated by the latest clang versions. It is still under heavy development but I'm working on it.

**Comment by guwirth** [comment]
@gyorb thanks for letting us know this. Looking forward to include a running solution.

**Comment by gyorb** [comment]
In the pull request there is an initial version which can import the main relevant parts of the reports.

There are some plans to improve this initial solution:
- [ ] use `issue_hash_content_of_line_in_contex` from the reports for uniqueing (generated based on the semantic context during the analysis)
- [ ] store report flow related messaged as in the [sonar-java plugin](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sslr%3Asslr&open=AVwbZ4h58FN7287hDv85&resolved=false&rules=squid%3AS2589) see discussion about the feature [here](https://groups.google.com/forum/?utm_source=digest&utm_medium=email/#!searchin/sonarqube/data$20flow%7Csort:relevance/sonarqube/9NknvkbkRZA/qmTVdWGoBQAJ), visualizing flow messages is available since SonarQube v6.4

Any feedback is appreciated.

**Comment by gyorb** [comment]
I've extended the wiki with some simple examples how to [analyze a project](https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Code-checkers#clang-static-analyzer) and with the [configuration option](https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Supported-configuration-properties#sonarcxxclangsareportpath) to import the plist reports.
